demonstrations/tutorial_pulse_programming101.py

```diff
@@ -0,0 +1,362 @@
+
+@qml.qnode(dev, interface="jax")
+def qnode(p, t=duration):
+    qml.BasisState(np.array([1, 1, 1, 0, 0, 0]), wires=H_obj.wires)
+    qml.evolve(H_pulse)(params=(*p, *p), t=t)
+    return qml.expval(H_obj)
+
+##############################################################################
+# Our aim is to minimize the energy expectation value :math:`\langle H_\text{obj} \rangle` but we also need to take 
+# certain physical constraints into account. For example, the amplitude of the physical driving field :math:`\Omega(t)` cannot 
+# reach arbitrarily large values and cannot change arbitrarily quickly in time. We therefore add two penalty terms quantifying the 
+# mean absolute square of the amplitude and changes in the amplitude. The overall cost function is thus
+# 
+# .. math:: \mathcal{C}(p) = \langle H_\text{obj} \rangle + \sum_{ki} |p^k_i|^2 + \sum_{ki} |p^k_i - p^k_{i-1}|^2.
+#
+
+def abs_diff(p):
+    """compute |p_i - p_i-1|^2"""
+    return jnp.mean(jnp.abs(jnp.diff(p[:, :-1], axis=1))**2)
+
+def cost_fn(params):
+    # params.shape = (n_wires, {n_bins}+1)
+    p = params[:, :-1]
+    C_exp = qnode(params)           # expectation value
+    C_par = jnp.mean(jnp.abs(p)**2) # parameter values
+    C_der = abs_diff(p)             # derivative values
+    return C_exp + 10*C_par + 10*C_der
+
+
+##############################################################################
+# We now have all the ingredients to run our ctrl-VQE program. We use the adam implementation in ``optax`` for optimizations in ``jax`` for our optimization loop.
+import optax 
+from datetime import datetime
+
+t_bins = 30 # number of time bins
+theta = jnp.array([jnp.ones(t_bins + 1, dtype=float) for _ in range(n_wires)])
+
+n_epochs = 100
+optimizer = optax.adam(learning_rate=0.1)
+opt_state = optimizer.init(theta)
+
+value_and_grad = jax.jit(jax.value_and_grad(cost_fn, argnums=0))
+
+energy = np.zeros(n_epochs)
+cost = np.zeros(n_epochs)
+theta_i = [theta]
+
+## Optimization loop
+
+time0 = datetime.now()
+_ = value_and_grad(theta)
+time1 = datetime.now()
+print(f"grad and val compilation time: {time1 - time0}")
+
+for n in range(n_epochs):
+    val, grad_circuit = value_and_grad(theta)
+    updates, opt_state = optimizer.update(grad_circuit, opt_state)
+    theta = optax.apply_updates(theta, updates)
+
+    energy[n] = qnode(theta)
+    cost[n] = val
+    theta_i.append(theta)
+
+    if not n%5:
+        print(f"{n+1} / {n_epochs}; energy: {energy[n]}; cost: {val}")
+
+##############################################################################
+# We see that the energy converges relatively quickly to the desired exact result.
+
+fig, axs = plt.subplots(nrows=2, figsize=(5,5), sharex=True)
+ax = axs[0]
+ax.plot(energy, ".:", label="energy")
+ax.plot([0, n_epochs], [E_exact]*2, ":", label="exact diag", color="grey")
+ax.set_ylabel("Energy")
+
+ax = axs[1]
+ax.plot(cost,".:", label="cost")
+ax.set_xlabel("epoch")
+ax.set_ylabel("Cost")
```

Comment:
Not sure about the labels, just a suggestion.
```suggestion
ax.plot(energy, ".:", label="$\langle H_\text{obj}\rangle (p)$")
ax.plot([0, n_epochs], [E_exact]*2, ":", label="$E_\text{min}$", color="grey")
ax.set_ylabel("Energy")
ax.legend()

ax = axs[1]
ax.plot(cost,".:", label="$C(p)$")
ax.set_xlabel("epoch")
ax.set_ylabel("Cost")
ax.legend()
```